### PR TITLE
[wip] Add spatial markov

### PR DESCRIPTION
**Pull request by ohasselblad**
This PR adds basic support for making trend predictions based on historic spatial measurements such as median income in census tracts.

Still need to:

- [x] ~~add predictions~~ (tabled for now)
- [x] add tests for all python functions
- [x] add tests for sql functions
- [x] documentation
- ~~[ ] add error handling~~ moved to another pull request TBD

### CODE REVIEW

**Comment by ohasselblad** [comment]
@jgoizueta, i'm testing this release on two different machines and getting values where are close but not equal, like this:

| local machine | remote server |
|---|---|
| 0.069767441860465115 | 0.0714285714286 |
| 0.065217391304347824 | 0.0666666666667 |

Which is fine given the precision of the measurements. How do you recommend testing situations like this in the testing environment? The diffs between what is produced on different machines doesn't match, but I could hard code some percent differences into the SQL [like here](https://github.com/CartoDB/observatory-extension/blob/develop/src/pg/test/sql/41_observatory_augmentation_test.sql#L224-L230) but I would like to expose the entire table to the tests..

**Comment by andrewxhill** [comment]
what's the current status here @ohasselblad ?

**Comment by ohasselblad** [comment]
merging today :)

**Comment by ohasselblad** [comment]
@stuartlynn, could you take a look?

File: doc/04_markov.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#### Returns
+
+A table with the following columns.
+
+| Column Name | Type | Description |
+|-------------|------|-------------|
+| trend | NUMERIC |  |
+| trend_up | NUMERIC |  |
+| trend_down | NUMERIC | The statistical significance (from 0 to 1) of a cluster or outlier classification. Lower numbers are more significant. |
```
Comment (stuartlynn) [inline doc/04_markov.md]: Is that description right? It sounds like its from Moran not Markov?

Comment (ohasselblad) [inline doc/04_markov.md]: good catch, giving docs some more love

File: src/pg/sql/11_markov.sql [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+--  2 | Pt2  | 11.0   | 13.2   | 12.5
+--  ...
+-- Sample Function call:
+-- SELECT CDB_SpatialMarkov('SELECT * FROM real_estate',
+--                          Array['date_1', 'date_2', 'date_3'])
+
+CREATE OR REPLACE FUNCTION
+  CDB_SpatialMarkovTrend (
+      subquery TEXT,
+      time_cols TEXT[],
```
Comment (stuartlynn) [inline src/pg/sql/11_markov.sql]: We could try using the [pyagg](https://github.com/CartoDB/crankshaft/blob/068c80f369dec4280cb1b6932aa6d57691836bbe/doc/04_pyAgg.md#pyagg-helper-function)  helper I wrote for Segmentation to collect and pass in the time columns and the data columns? Probably good to go as is but we could perhaps try and implement that as another function signature.

File: src/py/crankshaft/crankshaft/space_time_dynamics/markov.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    ## output the results
+    return zip(trend, trend_up, trend_down, volatility, weights.id_order)
+
+def get_time_data(markov_data, time_cols):
+    """
+        Extract the time columns and bin appropriately
+    """
+    num_attrs = len(time_cols)
+    return np.array([[x['attr' + str(i)] for x in markov_data]
```
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/space_time_dynamics/markov.py]: For speed it might be worth a quick refactor to use the array_agg trick to construct an array for each column in the query and then simply grab them by markov_data[0]['time_col_1'] etc. More of an enhancement than anything else though

**Comment by stuartlynn** [comment]
Left some comments inline but its looking good. Only thing that I think is missing is a way to somehow enforce equal interval timespans, the algorithm will give weird results if they aren't. Not sure how to do that just now with the approach of having different columns represent different timespans. 

Perhaps we need a time binning function that would generate the time bins from rows with a date column then pass them through to spatial markov function. So it would go something like : 

| time_stamp | value |
| --------------- | ------- |
| geom_1 | 2015-01-01| 2 |
| geom_2 | 2015-01-02| 3 |
| geom_3 | 2015-01-03| 3 |

To the table with columns for each day (including ones with no data) using a simple aggregation and then that goes in to the moran calculation? That way you could ensure the intervals are equal and allow prediction easily on different timescales. I think you have some of that internal though right?

**Comment by ohasselblad** [comment]
Yeah, maybe the most straight-forward way is to just tell people in the documentation that it expects the input time intervals to be equally spaced. 

A time-binning function would be pretty cool. That could be a step in the analysis pipeline. What do you think about putting it there instead of building more into this?

I built an internal tool for re-binning the time data using simple averages over the time bins, but @ljwolf thought that approach was a little to quick and dirty.

File: src/py/crankshaft/crankshaft/space_time_dynamics/markov.py (hunk 3 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/space_time_dynamics/markov.py]: 👍 will give it a shot

Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/space_time_dynamics/markov.py]: this is a pretty deep change that effects moran.py as well since the query-building is in pysal-utils. I'm going to create a ticket to make this change and do it in another PR. It's a needed change cause it's a pretty great speed up :)

**Comment by ljwolf** [comment]
Woah, this repo has moved quite a bit, thanks for the ping! I'm now watching the repo. 

IIRC, my concern about the averaging was mainly for uncertainty about the next-step prediction. When the data are temporally dependent, I think the expected value of 30 day's steps has different uncertainty than the resampled chain's single thirty-day step since, in general, the variance of a sum is not the sum of variances for dependent data. 

If the goal is to get a stable estimating chain over gaps, averaging seems pretty reasonable, but I wasn't sure what's statistically "correct" then, and got pulled off of this line of work by GSOC.

**Comment by rafatower** [comment]
Guys, as long as you are taking no responsibility in the release process I'm closing permissions to merge to master and develop branches. Sorry about that but this is getting a bit out of control.

I'll reopen it as soon as we can.

cc/ @javisantana

**Comment by ohasselblad** [comment]
@rafatower, we followed the guidelines listed here: https://github.com/CartoDB/crankshaft#working-process----quickstart-guide Specifically this line related to what we did for this pull request: "Developers shall create a new topic branch from develop for any new feature or bugfix and commit their changes to it and eventually merge back into the develop branch."

I'm confused about the development process again. The way I read that section is that we merge to develop and then ask for a release, which I did here: https://github.com/CartoDB/crankshaft/pull/71

**Comment by rafatower** [comment]
Just wanted to apologize cause my words led to some confusion:
- My intent was not to blame anyone at all. I do believe everyone followed what's documented as the process. I do believe that any deviations from the process (if any) are due to lack of proper specification.
- the process has been proved to be bad (or not as good as desired) and has a lot of overhead. I'm blaming the process, which I'm largely responsible for.
- I'm taking action to get the next release out ASAP, and
- I'll improve it to avoid unecessary overhead and risk in future releases

Excuse my poor communication skills if anyone felt offended. Please, please, let's keep the good work and also the good mood :sweat_smile:
